### PR TITLE
Implement some basic proposal validation (bsc#967684)

### PR DESCRIPTION
**Pull request by dirkmueller**
~~depends on https://github.com/SUSE-Cloud/automation/pull/879~~

### CODE REVIEW

File: crowbar_framework/app/models/trove_service.rb [hunk 1]
```diff
@@ -72,4 +72,20 @@ def proposal_dependencies(role)
...
   end
+
+  def validate_proposal_after_save(proposal)
+    validate_one_for_role proposal, "trove-server"
+
+    rabbitmq_proposal = Proposal.find_by(
+      barclamp: "rabbitmq",
+      name: proposal["attributes"]["rabbitmq"]["rabbitmq_instance"])
+
+    unless rabbitmq_proposal.nil?
```
Comment (vuntz) [inline crowbar_framework/app/models/trove_service.rb]: We should also fail if the rabbitmq proposal is nil, shouldn't we?

Comment (dirkmueller) [inline crowbar_framework/app/models/trove_service.rb]: probably yes, will do

**Comment by bmwiedemann** [comment]
retriggered testrun, after it correctly failed the validation

File: crowbar_framework/app/models/trove_service.rb [hunk 2]
```diff
@@ -72,4 +72,20 @@ def proposal_dependencies(role)
...
+    validate_one_for_role proposal, "trove-server"
+
+    rabbitmq_proposal = Proposal.find_by(
+      barclamp: "rabbitmq",
+      name: proposal["attributes"][@bc_name]["rabbitmq_instance"])
+
+    unless rabbitmq_proposal.nil?
+      unless rabbitmq_proposal["attributes"]["rabbitmq"]["trove"]["enabled"]
+        validation_error I18n.t(
+          "barclamp.#{@bc_name}.validation.rabbitmq_enabled")
```
Comment (bmwiedemann) [inline crowbar_framework/app/models/trove_service.rb]: -0.5 I dont think we really want to translate such technical JSON pathes.

Comment (MaximilianMeister) [inline crowbar_framework/app/models/trove_service.rb]: @bmwiedemann I think you are confusing this with the way we reference json trees in some of our tools. this is just referencing to https://github.com/crowbar/crowbar-openstack/pull/307/files#diff-6409fc1a00f53afd4af9f76751efb1dfR44

File: crowbar_framework/app/models/trove_service.rb [hunk 3]
```diff
@@ -72,4 +72,20 @@ def proposal_dependencies(role)
 
     answer
   end
+
+  def validate_proposal_after_save(proposal)
+    validate_one_for_role proposal, "trove-server"
```
Comment (MaximilianMeister) [inline crowbar_framework/app/models/trove_service.rb]: can we use parentheses here as well?

**Comment by MaximilianMeister** [comment]
> retriggered testrun, after it correctly failed the validation in https://ci.suse.de/job/openstack-mkcloud/17854/console
{"error":"Failed to validate proposal: translation missing: en.barclamp.trove.validation.rabbitmq_enabled\n"}


ah it seems this needs to go to the trove namespace in en.yml
`rabbitmq_enabled: 'Trove support has to be enabled in the RabbitMQ proposal first.`

File: crowbar_framework/app/models/trove_service.rb (hunk 3 above)
Comment (dirkmueller) [inline crowbar_framework/app/models/trove_service.rb]: do you mean validate_one_role proposal,"trove-server" ? probably we can, but houndci is not bitching about it, and it is consistent with all other invocations..

**Comment by dirkmueller** [comment]
@MaximilianMeister fixed

File: crowbar_framework/app/models/trove_service.rb (hunk 2 above)
Comment (dirkmueller) [inline crowbar_framework/app/models/trove_service.rb]: huh? thats a user visible error message.

**Comment by mjura** [comment]
+1

File: crowbar_framework/app/models/trove_service.rb (hunk 2 above)
Comment (bmwiedemann) [inline crowbar_framework/app/models/trove_service.rb]: It is a user-visible message, but how would the translated version look?
Currently it just throws an error about missing translation.

Comment (MaximilianMeister) [inline crowbar_framework/app/models/trove_service.rb]: @bmwiedemann that has been fixed, see comment below https://github.com/crowbar/crowbar-openstack/pull/307#issuecomment-189153426

File: crowbar_framework/app/models/trove_service.rb [hunk 4]
```diff
@@ -72,4 +72,20 @@ def proposal_dependencies(role)
...
+
+    rabbitmq_proposal = Proposal.find_by(
+      barclamp: "rabbitmq",
+      name: proposal["attributes"][@bc_name]["rabbitmq_instance"])
+
+    unless rabbitmq_proposal.nil?
```
Comment (vuntz) [inline crowbar_framework/app/models/trove_service.rb]: We're still missing an error if there's no rabbitmq proposal, aren't we?

Comment (vuntz) [inline crowbar_framework/app/models/trove_service.rb]: > We're still missing an error if there's no rabbitmq proposal, aren't we?

@dirkmueller ping

**Comment by tboerger** [comment]
:+1:

**Comment by toabctl** [comment]
+1

**Comment by toabctl** [comment]
@mjura still +1?

**Comment by mjura** [comment]
Yes, still +1

**Pull request merged**